Framework/Geometry/src/Instrument.cpp

```diff
@@ -1233,19 +1239,87 @@ const Beamline::DetectorInfo &Instrument::detectorInfo() const {
   return *m_detectorInfo;
 }
 
+/**
+ * @brief Instrument::hasComponentInfo
+ * @return True only if a ComponentInfo has been set
+ */
+bool Instrument::hasComponentInfo() const {
+  return static_cast<bool>(m_componentInfo);
+}
+
+/**
+ * @brief Instrument::componentInfo
+ * @return const ref to a ComponentInfo. Throws a std::runtime_error if
+ * not set.
+ */
+const Beamline::ComponentInfo &Instrument::componentInfo() const {
+  if (!hasComponentInfo()) {
+    throw std::runtime_error("Cannot return reference to NULL ComponentInfo");
+  }
+  return *m_componentInfo;
+}
+
+/**
+ * @brief Instrument::componentIds
+ * @return const ref to a vector of ComponentIds. Throws a std::runtime_error if
+ * Component Info not set.
+ */
+const std::vector<Geometry::ComponentID> &Instrument::componentIds() const {
+  if (!hasComponentInfo()) {
+    throw std::runtime_error(
+        "Cannot return component ids with a NULL ComponentInfo");
+  }
+  return m_componentIds;
+}
+
 /// Only for use by ExperimentInfo. Sets the pointer to the DetectorInfo.
 void Instrument::setDetectorInfo(
     boost::shared_ptr<const Beamline::DetectorInfo> detectorInfo) {
   m_detectorInfo = std::move(detectorInfo);
 }
 
+/**
+ * @brief Instrument::setComponentInfo
+ * @param componentInfo : ComponentInfo to store
+ * @param componentIds : ComponentIds to store
+ */
+void Instrument::setComponentInfo(
+    boost::shared_ptr<const Beamline::ComponentInfo> componentInfo,
+    std::vector<Geometry::ComponentID> componentIds) {
+  m_componentInfo = std::move(componentInfo);
+  m_componentIds = std::move(componentIds);
+}
+
 /// Returns the index for a detector ID. Used for accessing DetectorInfo.
 size_t Instrument::detectorIndex(const detid_t detID) const {
   const auto &baseInstr = m_map ? *m_instr : *this;
   const auto it = find(baseInstr.m_detectorCache, detID);
   return std::distance(baseInstr.m_detectorCache.cbegin(), it);
 }
 
+/**
+ * @brief Instrument::componentIndex
+ * Note that this index can be used with ComponentInfo. Throw std::runtime_error
+ * if the ComponentID does not exist.
+ * @param componentId : ComponentID to find the index for.
+ * @return the Component index associated with this Component.
+ */
+size_t Instrument::componentIndex(const ComponentID componentId) const {
```

Comment:
But `componentIndex()` is only called to access `ComponentInfo`, so if there is no `ComponentInfo`, `componentIndex()` will never get called? Or am I missing something here?